### PR TITLE
Implement general health checks in Api project (#1795)

### LINKED ISSUE #1795: Implement `Health Check` feature for the TodoTemplate project
It is required to have `Health Check` APIs for the template project.
These APIs can check the following items:
- available memory
- SQL server service
- SMTP service

### PR DESCRIPTION
**Pull request by milad-karimifard**
This closes #1795

### CODE REVIEW

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
                     {
-                        Reference = new OpenApiReference
-                        {
-                            Type = ReferenceType.SecurityScheme,
-                            Id = "bearerAuth"
-                        }
+                        Reference = new OpenApiReference { Type = ReferenceType.SecurityScheme, Id = "bearerAuth" }
                     },
-                    new string[] {}
+                    new string[] { }
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:106]: please roll back all of the code formatting changes above this line, because these changes are not related to this PR.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
     }
+
+    public static void AddHealthChecks(
+        this IServiceCollection services,
+        IConfiguration configuration,
+        IWebHostEnvironment environment)
+    {
+        var appsettings = configuration.GetSection(nameof(AppSettings)).Get<AppSettings>();
+
+        var healthCheckSetting = appsettings.HealCheckSettings;
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:119]: typo in the variable name, missing `s` at the end of `healthCheckSetting`.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+        
+        if (!healthCheckSetting.EnableHealthChecks)
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:121]: please do not use `!` operator at the start of a condition, it's not clear in the code. instead use a comparison with `false` like this:

```cs
if (healthCheckSettings.EnableHealthChecks is false)
```

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+            return;
+
+        var emailSettings = appsettings.EmailSettings;
+
+        services
+            .AddHealthChecksUI(setupSettings: setup =>
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:127]: please move this line of code up and put it after `services`

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+            {
+                setup.AddHealthCheckEndpoint("BitHealthCheck",
+                    environment.IsDevelopment() ? "http://localhost:5000/healthz" : "/healthz");
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:130]: is this check required? can't we just simply use `/healthz` for all of our environments?
@ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/appsettings.json [hunk 6]
```diff
@@ -24,17 +24,24 @@
...
-            "ResetPasswordEmailResendDelay": "0.00:02:00" //Format: D.HH:mm:nn
+            "ConfirmationEmailResendDelay": "0.00:02:00",
+            //Format: D.HH:mm:nn
+            "ResetPasswordEmailResendDelay": "0.00:02:00"
+            //Format: D.HH:mm:nn
         },
         "EmailSettings": {
-            "Host": "LocalFolder", // Local folder means storing emails in bin\sent-emails folder. (Recommended for testing purposes only)
+            "Host": "LocalFolder",
+            // Local folder means storing emails in bin\sent-emails folder. (Recommended for testing purposes only)
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/appsettings.json:34]: again, please roll back all of the code formatting changes above this line, because these changes are not related to this PR.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Startup/Middlewares.cs [hunk 7]
```diff
@@ -59,6 +61,11 @@ public static void Use(IApplicationBuilder app, IHostEnvironment env)
             app.UseEndpoints(endpoints =>
             {
                 endpoints.MapControllers().RequireAuthorization();
+                endpoints.MapHealthChecks("/healthz", new HealthCheckOptions
+                {
+                    Predicate = _ => true,
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Startup/Middlewares.cs:66]: isn't this line redundant?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 8]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+        services.AddHealthChecks()
+            .AddProcessAllocatedMemoryHealthCheck(healthCheckSetting.MaximumMegabytesAllocated)
+            .AddDiskStorageHealthCheck(opt =>
+                opt.AddDrive(Path.GetPathRoot(Directory.GetCurrentDirectory()), minimumFreeMegabytes: 5 * 1024))
+            .AddSqlServer(
+                configuration.GetConnectionString("SqlServerConnection"));
+
+        if (emailSettings.Host is not "LocalFolder")
+        {
+            services.AddHealthChecks()
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:143]: is it ok to call the `AddHealthChecks` method again here?
@ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+            {
+                setup.AddHealthCheckEndpoint("BitHealthCheck",
+                    environment.IsDevelopment() ? "http://localhost:5000/healthz" : "/healthz");
+            }).AddInMemoryStorage();
+
+
+        services.AddHealthChecks()
+            .AddProcessAllocatedMemoryHealthCheck(healthCheckSetting.MaximumMegabytesAllocated)
+            .AddDiskStorageHealthCheck(opt =>
+                opt.AddDrive(Path.GetPathRoot(Directory.GetCurrentDirectory()), minimumFreeMegabytes: 5 * 1024))
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:137]: isn't it better to put the hard-coded value of `5 * 1024` into the settings?
@ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+            .AddSqlServer(
+                configuration.GetConnectionString("SqlServerConnection"));
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:139]: hmm, I don't like this hard-coded string here (`"SqlServerConnection"`) 🤔
@ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 11]
```diff
@@ -89,30 +85,72 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+
+        if (emailSettings.Host is not "LocalFolder")
```
Comment (msynk) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:141]: again, hard-coded string (`"LocalFolder"`), can't we do something for these values?
@ysmoradi

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/AppSettings.cs [hunk 12]
```diff
@@ -8,11 +8,18 @@ public class AppSettings
 
         public EmailSettings EmailSettings { get; set; }
 
+        public HealCheckSettings HealCheckSettings { get; set; }
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/AppSettings.cs:11]: Please put new line after this

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/AppSettings.cs [hunk 13]
```diff
@@ -8,11 +8,18 @@ public class AppSettings
...
         public string UserProfileImagePath { get; set; }
 
         public string WebServerAddress { get; set; }
     }
 
+    public class HealCheckSettings
+    {
+        public bool EnableHealthChecks { get; set; }
+        public int MaximumMegabytesAllocated { get; set; }
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/AppSettings.cs:20]: What's this maximum megabytes...?

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs (hunk 9 above)
Comment (ysmoradi) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:137]: I thinks it's fine to use hard coded values
But the value itself is too small
I'd recommend 6 * 1024 * 1024

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/TodoTemplate.Api.csproj [hunk 14]
```diff
@@ -34,6 +34,12 @@
         </PackageReference>
         <PackageReference Include="MimeTypeMapOfficial" Version="1.0.17" />
         <PackageReference Include="Swashbuckle.AspNetCore" Version="6.3.1" />
+        <PackageReference Include="AspNetCore.HealthChecks.SqlServer" Version="6.0.2" />
+        <PackageReference Include="AspNetCore.HealthChecks.System" Version="6.0.4" />
+        <PackageReference Include="AspNetCore.HealthChecks.UI" Version="6.0.4" />
+        <PackageReference Include="AspNetCore.HealthChecks.UI.Client" Version="6.0.4" />
+        <PackageReference Include="AspNetCore.HealthChecks.UI.InMemory.Storage" Version="6.0.4" />
+        <PackageReference Include="AspNetCore.HealthChecks.Network" Version="6.0.3" />
     </ItemGroup>
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/TodoTemplate.Api.csproj:43]: Do we need networking package?

**Review by ysmoradi — CHANGES_REQUESTED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs (hunk 9 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:137]: hmmm, since there is a setting introduced to store `MaximumMegabytesAllocated` in this PR, we need to decide which way is better, using the hard-coded values or using the settings. I'm ok with either way, but not the current mix & match 😁
@ysmoradi

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs (hunk 5 above)
Comment (milad-karimifard) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:130]: I tested that and got an exception, but now tried again and it's working. I assume this because of misconfiguration at that moment. I fix it in next commit

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs (hunk 9 above)
Comment (milad-karimifard) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:137]: I think it better to let the end-user decide about their application memory usage

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/TodoTemplate.Api.csproj (hunk 14 above)
Comment (milad-karimifard) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/TodoTemplate.Api.csproj:43]: It's for Smtp health check

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs (hunk 5 above)
Comment (msynk) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:130]: That's great, so please remove all of the `environment` argument passing between methods that you've implemented too.

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs (hunk 8 above)
Comment (milad-karimifard) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:146]: When calling AddHealthChecks it's try to register itself and finally returns IHealthCheckBuilder so I think is fine

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/AppSettings.cs (hunk 13 above)
Comment (milad-karimifard) [reply src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/AppSettings.cs]: I renamed it to MaximumApplicationMemoryMegabytesAllocated. I think it is better to let the end-user decide about their application memory usage. also, 6 * 1024 * 1024 means 6 TB and it's very huge because unit it's megabyte

**Comment by msynk** [comment]
@ysmoradi I think this PR is good enough for now.
please post your response & opinion about the comments I've mentioned you in. and then we can approve and merge this PR.

**Review by msynk — APPROVED** [review]

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 15]
```diff
@@ -115,4 +115,44 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+    {
+        var appsettings = configuration.GetSection(nameof(AppSettings)).Get<AppSettings>();
+
+        var healthCheckSettings = appsettings.HealCheckSettings;
+        
+        if (healthCheckSettings.EnableHealthChecks is false)
+            return;
+        
+        services.AddHealthChecksUI(setupSettings: setup =>
+            {
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:129]: Seems it has extra tab space

File: src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs [hunk 16]
```diff
@@ -115,4 +115,44 @@ public static void AddTodoTemplateSwaggerGen(this IServiceCollection services)
...
+                setup.AddHealthCheckEndpoint("BitHealthCheck", "/healthz");
+            }).AddInMemoryStorage();
+
+
```
Comment (ysmoradi) [inline src/Tooling/Bit.Tooling.Templates/TodoTemplate/Api/Extensions/IServiceCollectionExtensions.cs:133]: Extra new line?

**Review by ysmoradi — APPROVED** [review]

**Review by ysmoradi — CHANGES_REQUESTED** [review]
Could you please add health check for efcore dbcontext instead of sql connection string?
AddDbContextCheck

Also check this on wsl too

**Comment by milad-karimifard** [comment]
> 
@ysmoradi 
I tested on `wsl` and it works like a champ

**Review by msynk — APPROVED** [review]

**Review by ysmoradi — APPROVED** [review]
